compiler/modules/front.syntax/token.bal

```diff
@@ -499,6 +528,57 @@ public readonly class SourceFile {
     }
 }
 
+function tokenEndCodePointIndex(string[] fragments, FragCode[] fragCodes, int startCodePointIndex) returns int {
+    int fragmentIndex = fragmentCountUpTo(fragments, startCodePointIndex);
+    return lineTokenEndIndex(fragments, fragCodes, startCodePointIndex, fragmentIndex);
+}
+
+function fragmentCountUpTo(string[] fragments, int startCodePointIndex) returns int {
+    int nCodePoints = 0;
+    int fragmentIndex = 0;
+    int nFragments = fragments.length();
+    while fragmentIndex < nFragments {
+        if nCodePoints >= startCodePointIndex {
+            break;
+        }
+        nCodePoints += fragments[fragmentIndex].length();
+        fragmentIndex += 1;
+    }
+    return fragmentIndex;
+}
+
+// Return code-point index of end of token that starts in
+function lineTokenEndIndex(string[] fragments, FragCode[] fragCodes, int startCodePointIndex, int startFragmentIndex) returns int {
+    match fragCodes[startFragmentIndex] {
+        FRAG_STRING_OPEN  => {
+            return stringTokenEndIndex(fragments, fragCodes, startCodePointIndex, startFragmentIndex);
+        }
+        FRAG_GREATER_THAN => {
+            // Assume not in type-desc mode
+            if startFragmentIndex + 1 < fragCodes.length() && fragCodes[startFragmentIndex+1] == FRAG_GREATER_THAN {
+                if startFragmentIndex + 2 < fragCodes.length() && fragCodes[startFragmentIndex+2] == FRAG_GREATER_THAN {
+                    return startCodePointIndex + 3; // >>>
+                }
+                return startCodePointIndex + 2; // >>
+            }
+            return startCodePointIndex + 1; // >
+        }
+    }
+    return startCodePointIndex + fragments[startFragmentIndex].length();
+}
+
+function stringTokenEndIndex(string[] fragments, FragCode[] fragCodes, int startCodePointIndex, int startFragmentIndex) returns int {
```

Comment:
Call this `stringTokenEndCodePointIndex`.

Reply:
fixed